### PR TITLE
Webhook high load scenario

### PR DESCRIPTION
**Pull request by alex-ko-dev**

### CODE REVIEW

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    private final List<CompletableFuture<Long>> requestSendingTasks = new ArrayList<>();
+    private final GitHubProperties properties;
+
+    public WebHookSteps(GitHubProperties properties) {
+        this.properties = properties;
+    }
+
+    @Given("CxFlow is running as a service")
+    public void runAsService() {
+        SpringApplication.run(CxFlowApplication.class, "--web");
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: this should be moved to a utility, we are starting to use it in too many places

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @When("GitHub sends WebHook requests to CxFlow {int} times per second")
+    public void githubSendsWebHookRequests(int timesPerSecond) {
+        final int MAX_TOTAL_REQUESTS = 10,
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: better to move magic numbers to property file

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                MILLISECONDS_IN_SECOND = 1000;
+
+        int intervalBetweenRequestsMs = MILLISECONDS_IN_SECOND / timesPerSecond;
+
+        logger.info("Starting to send WebHook requests with the interval of {} ms.", intervalBetweenRequestsMs);
+        for (int i = 0; i < MAX_TOTAL_REQUESTS; i++) {
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: there are better implementations for doing things in time intervals than a for loop.

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    private void sendWebHookRequest() {
+        throw new PendingException();
+    }
+
+    private void waitForAllTasksToComplete(List<CompletableFuture<Long>> tasks) {
+        logger.info("Waiting for all the requests to complete.");
+        CompletableFuture[] futureArray = tasks.toArray(new CompletableFuture[0]);
+        CompletableFuture<Void> combinedTask = CompletableFuture.allOf(futureArray);
+        Assert.assertFalse(combinedTask.isCompletedExceptionally());
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: please add message to the assertion

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Then("each of the requests is answered in at most {int} ms")
+    public void eachOfTheRequestsIsAnsweredInAtMostMs(long expectedMaxDurationMs) {
+        Optional<Long> actualMaxDurationMs = requestSendingTasks.stream()
+                .map(WebHookSteps::toExecutionTimeMs)
+                .max(Long::compare);
+
+        Assert.assertTrue(actualMaxDurationMs.isPresent());
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: please add message to assertion

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    public void runAsService() {
+        SpringApplication.run(CxFlowApplication.class, "--web");
+    }
+
+    @When("GitHub sends WebHook requests to CxFlow {int} times per second")
+    public void githubSendsWebHookRequests(int timesPerSecond) {
+        final int MAX_TOTAL_REQUESTS = 10,
+                MILLISECONDS_IN_SECOND = 1000;
+
+        int intervalBetweenRequestsMs = MILLISECONDS_IN_SECOND / timesPerSecond;
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: is int a good choise here???

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+        String message = String.format("Actual duration (%d ms) is greater than the expected max duration (%d ms).",
+                actualMaxDurationMs.get(),
+                expectedMaxDurationMs);
+
+        Assert.assertTrue(message, actualMaxDurationMs.get() <= expectedMaxDurationMs);
+    }
+
+    private static long toExecutionTimeMs(CompletableFuture<Long> task) {
+        try {
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: what about the getexceptionaly()?

File: src/test/resources/cucumber/features/componentTests/webhook.feature [hunk 8]
```diff
@@ -1,10 +1,9 @@
...
+@WebHookFeature @ComponentTest
 Feature: Check Component tests Webhook functionality
 
   Scenario: High load on CxFlow due to frequent WebHook requests from GitHub
-    Given source is GitHub
-    And CxFlow is running as a service
+    Given version control provider is GitHub
+    Given CxFlow is running as a service
     And webhook is configured for push event
     When GitHub sends WebHook requests to CxFlow 2 times per second
```
Comment (ofersk) [inline src/test/resources/cucumber/features/componentTests/webhook.feature]: should this be a parameter?

File: src/test/resources/cucumber/features/componentTests/webhook.feature [hunk 9]
```diff
@@ -1,10 +1,9 @@
...
-    Then 95% of the requests are answered in 200 ms or less
-    And the rest 5% of the requests are answered in between 200 ms and 500 ms
\ No newline at end of file
+    Then each of the requests is answered in at most 500 ms
```
Comment (ofersk) [inline src/test/resources/cucumber/features/componentTests/webhook.feature]: should we test benchmark tests with cucumber?
this way we ignore all the statistical fluctuations and other important aspects 👎

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java [hunk 10]
```diff
@@ -2,105 +2,177 @@
...
+    private GitHubProperties properties;
+
+    private HttpEntity<String> webHookRequest;
+    private String cxFlowPort;
 
     @Given("CxFlow is running as a service")
     public void runAsService() {
-        SpringApplication.run(CxFlowApplication.class, "--web");
+        ConfigurableApplicationContext context = SpringApplication.run(CxFlowApplication.class, "--web");
+        cxFlowPort = context.getEnvironment().getProperty("server.port");
```
Comment (ofersk) [inline src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: run as a service logic should be exported to a utility

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java (hunk 7 above)
Comment (alex-ko-dev) [reply src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: Do you mean isCompletedExceptionally()? Even if I add it as a condition, I'll still need to use try..catch, so checking for isCompletedExceptionally() seems unnecessary here.

File: src/test/resources/cucumber/features/componentTests/webhook.feature (hunk 8 above)
Comment (alex-ko-dev) [reply src/test/resources/cucumber/features/componentTests/webhook.feature]: It's already a parameter:
```
@When("GitHub sends WebHook requests to CxFlow {int} times per second")
'public void githubSendsWebHookRequests(int timesPerSecond) {
```

File: src/test/resources/cucumber/features/componentTests/webhook.feature (hunk 9 above)
Comment (alex-ko-dev) [reply src/test/resources/cucumber/features/componentTests/webhook.feature]: This is basically a compromise between cucumber readability, code simplicity vs more advanced benchmark testing tools.
I suggest that we stay with cucumber for now.

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java (hunk 4 above)
Comment (alex-ko-dev) [reply src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: I've noticed that this check is already covered by a null check further in code, so I've removed this assertion,

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java (hunk 6 above)
Comment (alex-ko-dev) [reply src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: Changed to Duration. 
Regarding int vs long, we don't expect to work with billions in this case :)

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java (hunk 10 above)
Comment (alex-ko-dev) [reply src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: Done: TestUtils.runCxFlowAsService()

File: src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java (hunk 3 above)
Comment (alex-ko-dev) [reply src/test/java/com/checkmarx/flow/cucumber/component/webhook/WebHookSteps.java]: I could use Timer or ScheduledExecutorService, but they don't support specifying the number of executions, so I'll need to implement some kind of a shared variable in addition to Futures, which will complicate the code too much.

**Pull request merged**